rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -43,47 +43,116 @@
 namespace rosbag2_transport
 {
 
-Recorder::Recorder(
-  const std::string & node_name,
-  const rclcpp::NodeOptions & node_options)
-: rclcpp::Node(node_name, node_options)
+class RecorderImpl
 {
-  // TODO(karsten1987): Use this constructor later with parameter parsing.
-  // The reader, storage_options as well as record_options can be loaded via parameter.
-  // That way, the recorder can be used as a simple component in a component manager.
-  throw rclcpp::exceptions::UnimplementedError();
-}
-
-Recorder::Recorder(
-  std::shared_ptr<rosbag2_cpp::Writer> writer,
-  const rosbag2_storage::StorageOptions & storage_options,
-  const rosbag2_transport::RecordOptions & record_options,
-  const std::string & node_name,
-  const rclcpp::NodeOptions & node_options)
-: Recorder(
-    std::move(writer),
-#ifndef _WIN32
-    std::make_shared<KeyboardHandler>(false),
-#else
-    // We don't have signal handler option in constructor for windows version
-    std::shared_ptr<KeyboardHandler>(new KeyboardHandler()),
-#endif
-    storage_options,
-    record_options,
-    node_name,
-    node_options)
-{}
-
-Recorder::Recorder(
+public:
+  RecorderImpl(
+    rclcpp::Node * owner,
+    std::shared_ptr<rosbag2_cpp::Writer> writer,
+    std::shared_ptr<KeyboardHandler> keyboard_handler,
+    const rosbag2_storage::StorageOptions & storage_options,
+    const rosbag2_transport::RecordOptions & record_options);
+
+  ~RecorderImpl();
+
+  //////////////////////////////////////
+  // Public/protected interface mirrors
+
+  void record();
+
+  const rosbag2_cpp::Writer & get_writer_handle();
+
+  /// Pause the recording.
+  void pause();
+
+  /// Resume recording.
+  void resume();
+
+  /// Pause if it was recording, continue recording if paused.
+  void toggle_paused();
+
+  /// Return the current paused state.
+  bool is_paused();
+
+  std::unordered_map<std::string, std::string> get_requested_or_available_topics();
+
+  ////////////////////////////
+  // Implementation helpers
+
+  void topics_discovery();
```

Comment:
@emersonknapp It seems almost all `"Implementation helpers"` could be private.
Could you please make them private?

Reply:
done - left a few members public to the wrapper to avoid extra unecessary getters